server/src/main/java/org/elasticsearch/index/query/CombinedFieldsQueryBuilder.java

```diff
@@ -0,0 +1,433 @@
+            } else {
+                throw new IllegalArgumentException("Unsupported [" + ZERO_TERMS_QUERY_FIELD.getPreferredName() + "] value [" + value + "]");
+            }
+        }, ZERO_TERMS_QUERY_FIELD);
+
+        PARSER.declareFloat(CombinedFieldsQueryBuilder::boost, BOOST_FIELD);
+        PARSER.declareString(CombinedFieldsQueryBuilder::queryName, NAME_FIELD);
+    }
+
+    /**
+     * Constructs a new text query.
+     */
+    public CombinedFieldsQueryBuilder(Object value, String... fields) {
+        if (value == null) {
+            throw new IllegalArgumentException("[" + NAME + "] requires query value");
+        }
+        if (fields == null) {
+            throw new IllegalArgumentException("[" + NAME + "] requires field list");
+        }
+        this.value = value;
+        this.fieldsAndBoosts = new TreeMap<>();
+        for (String field : fields) {
+            field(field);
+        }
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public CombinedFieldsQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        value = in.readGenericValue();
+        int size = in.readVInt();
+        fieldsAndBoosts = new TreeMap<>();
+        for (int i = 0; i < size; i++) {
+            String field = in.readString();
+            float boost = in.readFloat();
+            checkNegativeBoost(boost);
+            fieldsAndBoosts.put(field, boost);
+        }
+        operator = Operator.readFromStream(in);
+        minimumShouldMatch = in.readOptionalString();
+        zeroTermsQuery = ZeroTermsQueryOption.readFromStream(in);
+        autoGenerateSynonymsPhraseQuery = in.readBoolean();
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeGenericValue(value);
+        out.writeVInt(fieldsAndBoosts.size());
+        for (Map.Entry<String, Float> fieldsEntry : fieldsAndBoosts.entrySet()) {
+            out.writeString(fieldsEntry.getKey());
+            out.writeFloat(fieldsEntry.getValue());
+        }
+        operator.writeTo(out);
+        out.writeOptionalString(minimumShouldMatch);
+        zeroTermsQuery.writeTo(out);
+        out.writeBoolean(autoGenerateSynonymsPhraseQuery);
+    }
+
+    public Object value() {
+        return value;
+    }
+
+    /**
+     * Adds a field to run the query against.
+     */
+    public CombinedFieldsQueryBuilder field(String field) {
+        if (Strings.isEmpty(field)) {
+            throw new IllegalArgumentException("supplied field is null or empty.");
+        }
+        this.fieldsAndBoosts.put(field, AbstractQueryBuilder.DEFAULT_BOOST);
+        return this;
+    }
+
+    /**
+     * Adds a field to run the query against with a specific boost.
+     */
+    public CombinedFieldsQueryBuilder field(String field, float boost) {
+        if (Strings.isEmpty(field)) {
+            throw new IllegalArgumentException("supplied field is null or empty.");
+        }
+        checkNegativeBoost(boost);
```

Comment:
do we also want check for  `boost < 1` similar to other places?

Reply:
Yes I think we should! I had also forgotten to check in the `fields` method. I added a new test case for this, `CombinedFieldsQueryParsingTests#testInvalidFieldBoosts`.